client/me/form-base/with-form-base.jsx

```diff
@@ -8,96 +8,54 @@ import { connect } from 'react-redux';
+import { isUpdatingUserSettings } from 'calypso/state/user-settings/selectors';
 
 const withFormBase = ( WrappedComponent ) => {
 	class EnhancedComponent extends React.Component {
 		static displayName = `withFormBase(${ WrappedComponent.displayName || WrappedComponent.name })`;
 
 		state = {
 			redirect: false,
-			submittingForm: false,
-			showNotice: false,
 		};
 
-		componentDidMount() {
-			userSettings.getSettings();
-		}
-
 		componentWillUnmount() {
 			// Silently clean up unsavedSettings before unmounting
-			userSettings.unsavedSettings = {};
-		}
-
-		componentDidUpdate() {
-			this.showNotice();
+			this.props.clearUnsavedUserSettings();
 		}
 
 		getDisabledState = () => {
-			return this.state.submittingForm;
+			return this.props.isUpdatingUserSettings;
 		};
 
-		showNotice() {
-			if ( userSettings.initialized && this.state.showNotice ) {
-				this.props.successNotice( this.props.translate( 'Settings saved successfully!' ), {
-					id: 'form-base',
-				} );
-				this.state.showNotice = false;
-			}
-		}
-
 		getSetting = ( settingName ) => {
-			return userSettings.getSetting( settingName ) || '';
+			const { _unsavedUserSettings, _userSettings } = this.props;
+			return _unsavedUserSettings[ settingName ] ?? _userSettings[ settingName ] ?? '';
 		};
 
 		toggleSetting = ( event ) => {
 			const { name } = event.currentTarget;
-			userSettings.updateSetting( name, ! this.getSetting( name ) );
+			this.props.setUserSetting( name, ! this.props._userSettings[ name ] );
 		};
 
 		updateSetting = ( event ) => {
 			const { name, value } = event.currentTarget;
-			userSettings.updateSetting( name, value );
+			this.props.setUserSetting( name, value );
 		};
 
 		submitForm = ( event ) => {
 			event.preventDefault();
 
-			this.setState( { submittingForm: true } );
-			userSettings.saveSettings(
-				function ( error ) {
-					if ( error ) {
-						// handle error case here
-						if ( error.message ) {
-							this.props.errorNotice( error.message, { id: 'form-base' } );
-						} else {
-							this.props.errorNotice(
-								this.props.translate( 'There was a problem saving your changes.' ),
-								{ id: 'form-base' }
-							);
-						}
-						this.setState( { submittingForm: false } );
-					} else {
-						this.props.markSaved && this.props.markSaved();
-
-						if ( this.state && this.state.redirect ) {
-							user()
-								.clear()
-								.then( () => {
-									// Sometimes changes in settings require a url refresh to update the UI.
-									// For example when the user changes the language.
-									window.location = this.state.redirect + '?updated=success';
-								} );
-							return;
-						}
-						// if we set submittingForm too soon the UI updates before the response is handled
-						this.setState( { showNotice: true, submittingForm: false } );
-						this.showNotice();
-					}
-				}.bind( this )
-			);
+			this.props.saveUserSettings( null, () => this.props.markSaved?.() );
```

Comment:
This `onSuccess` pattern is a bit odd - almost like an attempt to mirror the promise functionality in Redux. Any alternatives or thoughts?